File: src/renderer/vlmClient.ts
import type { PlanSuggestion } from '../shared/contracts';
import { parsePlanSuggestion } from '../shared/planParser';

const SYSTEM_PROMPT = `You are a visual planning assistant for a browser screenshot.
Return exactly one JSON object and no surrounding commentary.

Required JSON schema:
{
  "summary": string,
  "current_view": string,
  "goals": string[],
  "next_steps": string[],
  "risks": string[],
  "confidence": number
}

Rules:
- Keep the response concise.
- confidence must be a number between 0 and 1.
- If something is unclear, mention it in risks instead of inventing certainty.
- Do not wrap the JSON in markdown fences.`;

const MOCK_PLAN: PlanSuggestion = {
  summary:
    'The screenshot shows the browser-based Poker VLM interface ready to analyze a selected tab or window.',
  current_view: 'Browser baseline home screen',
  goals: [
    'Capture the selected browser surface',
    'Send the image to the configured VLM endpoint',
  ],
  next_steps: [
    'Confirm the captured preview matches the intended tab or window',
    'Use the debug JSON panel to refine the prompt or downstream schema',
  ],
  risks: [
    'The browser capture picker requires manual user approval on each request',
    'Mock mode is enabled, so no live VLM call has been made yet',
  ],
  confidence: 0.51,
};

type AppConfig = {
  baseUrl: string;
  apiKey?: string;
  model: string;
  timeoutMs: number;
};

type ChatCompletionResponse = {
  choices?: Array<{
    message?: {
      content?:
        | string
        | Array<{
            text?: string;
            type?: string;
          }>;
    };
  }>;
};

function getAppConfig(): AppConfig {
  const timeout = Number.parseInt(import.meta.env.VITE_VLM_TIMEOUT_MS ?? '20000', 10);

  return {
    baseUrl: import.meta.env.VITE_VLM_BASE_URL?.trim() || 'mock',
    apiKey: import.meta.env.VITE_VLM_API_KEY?.trim() || undefined,
    model: import.meta.env.VITE_VLM_MODEL?.trim() || 'gpt-4.1-mini',
    timeoutMs: Number.isFinite(timeout) && timeout > 0 ? timeout : 20_000,
  };
}

function extractMessageText(response: ChatCompletionResponse): string {
  const content = response.choices?.[0]?.message?.content;

  if (typeof content === 'string' && content.trim().length > 0) {
    return content;
  }

  if (Array.isArray(content)) {
    const text = content
      .map((entry) => entry.text?.trim())
      .filter((entry): entry is string => Boolean(entry))
      .join('\n')
      .trim();

    if (text.length > 0) {
      return text;
    }
  }

  throw {
    code: 'VLM_EMPTY_RESPONSE',
    message: 'The VLM response did not contain any text content.',
  };
}

export async function requestPlanSuggestion(imageDataUrl: string): Promise<{
  plan: PlanSuggestion;
  rawResponse: string;
}> {
  const config = getAppConfig();

  if (config.baseUrl === 'mock') {
    return {
      plan: MOCK_PLAN,
      rawResponse: JSON.stringify(MOCK_PLAN, null, 2),
    };
  }

  const controller = new AbortController();
  const timeoutId = window.setTimeout(() => controller.abort(), config.timeoutMs);

  try {
    const response = await fetch(
      `${config.baseUrl.replace(/\/$/, '')}/chat/completions`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...(config.apiKey
            ? { Authorization: `Bearer ${config.apiKey}` }
            : {}),
        },
        body: JSON.stringify({
          model: config.model,
          temperature: 0,
          messages: [
            {
              role: 'system',
              content: SYSTEM_PROMPT,
            },
            {
              role: 'user',
              content: [
                {
                  type: 'text',
                  text:
                    'Analyze this browser screenshot and return the requested JSON plan. Focus on what the user should do next.',
                },
                {
                  type: 'image_url',
                  image_url: {
                    url: imageDataUrl,
                  },
                },
              ],
            },
          ],
        }),
        signal: controller.signal,
      },
    );

    const responseText = await response.text();

    if (!response.ok) {
      throw {
        code: 'VLM_HTTP_ERROR',
        message: `The VLM request failed with ${response.status} ${response.statusText}.`,
        rawResponse: responseText,
      };
    }

    let parsedResponse: ChatCompletionResponse;
    try {
      parsedResponse = JSON.parse(responseText) as ChatCompletionResponse;
    } catch {
      throw {
        code: 'VLM_BAD_JSON',
        message: 'The VLM returned invalid JSON.',
        rawResponse: responseText,
      };
    }

    const rawResponse = extractMessageText(parsedResponse);

    try {
      return {
        plan: parsePlanSuggestion(rawResponse),
        rawResponse,
      };
    } catch (error) {
      throw {
        code: 'VLM_SCHEMA_ERROR',
        message: 'The VLM response did not match the expected plan schema.',
        details: error instanceof Error ? error.message : undefined,
        rawResponse,
      };
    }
  } catch (error) {
    if (error instanceof DOMException && error.name === 'AbortError') {
      throw {
        code: 'VLM_TIMEOUT',
        message: `The VLM request timed out after ${config.timeoutMs}ms.`,
      };
    }

    if (error instanceof TypeError) {
      throw {
        code: 'VLM_NETWORK_ERROR',
        message:
          'The browser could not reach the VLM endpoint. Check the URL, make sure the server is running, and verify CORS is enabled for browser requests.',
      };
    }

    throw error;
  } finally {
    window.clearTimeout(timeoutId);
  }
}

File: app/vlm_client.py
from __future__ import annotations

from typing import Any

import httpx

from app.config import get_config
from app.errors import AppError
from app.models import PlanSuggestion
from app.parser import parse_plan_suggestion

SYSTEM_PROMPT = """You are a visual planning assistant for a browser screenshot.
Return exactly one JSON object and no surrounding commentary.

Required JSON schema:
{
  "summary": string,
  "current_view": string,
  "goals": string[],
  "next_steps": string[],
  "risks": string[],
  "confidence": number
}

Rules:
- Keep the response concise.
- confidence must be a number between 0 and 1.
- If something is unclear, mention it in risks instead of inventing certainty.
- Do not wrap the JSON in markdown fences."""

MOCK_PLAN = PlanSuggestion(
    summary=(
        "The screenshot shows the Python-based OpenFish interface ready to analyze a selected monitor, window, or tab."
    ),
    current_view="OpenFish Python baseline home screen",
    goals=[
        "Capture the selected browser surface",
        "Send the screenshot through the Python VLM client",
    ],
    next_steps=[
        "Confirm the preview matches the intended source",
        "Review the structured plan before extending the workflow",
    ],
    risks=[
        "Mock mode is enabled, so no live VLM call has been made",
        "This branch keeps browser capture in JavaScript because web capture requires client APIs",
    ],
    confidence=0.57,
)


def _extract_message_text(payload: dict[str, Any]) -> str:
    choices = payload.get("choices") or []
    if not choices:
        raise AppError(
            "VLM_EMPTY_RESPONSE",
            "The VLM response did not contain any choices.",
            status_code=502,
        )

    content = choices[0].get("message", {}).get("content")

    if isinstance(content, str) and content.strip():
        return content.strip()

    if isinstance(content, list):
        text_parts = []
        for item in content:
            text = item.get("text")
            if isinstance(text, str) and text.strip():
                text_parts.append(text.strip())
        if text_parts:
            return "\n".join(text_parts)

    raise AppError(
        "VLM_EMPTY_RESPONSE",
        "The VLM response did not contain text content.",
        status_code=502,
    )


async def request_plan_suggestion(image_data_url: str) -> tuple[PlanSuggestion, str]:
    config = get_config()

    if config.base_url == "mock":
        raw = MOCK_PLAN.model_dump_json(indent=2)
        return MOCK_PLAN, raw

    headers = {"Content-Type": "application/json"}
    if config.api_key:
        headers["Authorization"] = f"Bearer {config.api_key}"

    payload = {
        "model": config.model,
        "temperature": 0,
        "messages": [
            {"role": "system", "content": SYSTEM_PROMPT},
            {
                "role": "user",
                "content": [
                    {
                        "type": "text",
                        "text": (
                            "Analyze this browser screenshot and return the requested "
                            "JSON plan. Focus on what the user should do next."
                        ),
                    },
                    {
                        "type": "image_url",
                        "image_url": {"url": image_data_url},
                    },
                ],
            },
        ],
    }

    try:
        async with httpx.AsyncClient(timeout=config.timeout_ms / 1000) as client:
            response = await client.post(
                f"{config.base_url.rstrip('/')}/chat/completions",
                json=payload,
                headers=headers,
            )
    except httpx.TimeoutException as exc:
        raise AppError(
            "VLM_TIMEOUT",
            f"The VLM request timed out after {config.timeout_ms}ms.",
            status_code=504,
        ) from exc
    except httpx.HTTPError as exc:
        raise AppError(
            "VLM_NETWORK_ERROR",
            "The Python backend could not reach the VLM endpoint.",
            details=str(exc),
            status_code=502,
        ) from exc

    response_text = response.text
    if response.status_code >= 400:
        raise AppError(
            "VLM_HTTP_ERROR",
            f"The VLM request failed with {response.status_code} {response.reason_phrase}.",
            raw_response=response_text,
            status_code=502,
        )

    try:
        parsed = response.json()
    except ValueError as exc:
        raise AppError(
            "VLM_BAD_JSON",
            "The VLM returned invalid JSON.",
            raw_response=response_text,
            status_code=502,
        ) from exc

    raw_message = _extract_message_text(parsed)
    plan = parse_plan_suggestion(raw_message)
    return plan, raw_message
